### ISSUE TITLE
Something with scale

### ISSUE THREAD

**Issue by munrocket** [ilastik/marching_cubes#25]
Hello! Your library is incredibly fast and with good volume interpolation.
But why I get such a weird glitch with scale? Any thoughts?

![](https://habrastorage.org/webt/pw/2c/c9/pw2cc9jdxfjqltxl93jtysrat9s.png)

**Comment by stuarteberg** [comment]
I’m not sure I understand the problem.  Are you saying that your object appears abnormally elongated along one of its axes? Can you post the code and perhaps a test file to reproduce the issue?

**Comment by munrocket** [comment]
@stuarteberg yep, exactly.

[Here](https://drive.google.com/file/d/1fyQBrc3699YRfmczdbSF936LqP9oMkgJ/view?usp=sharing) I saved my tensor with `torch.save(tensor, 'torch_tensor.pt')`

My code
```python
import numpy as np
import torch
from stl import mesh
from marching_cubes import march
#import mcubes

mask = torch.load('torch_tensor.pt')

verts, normals, faces = march(mask.numpy(), 4)
#verts, faces = mcubes.marching_cubes(mask, 0)

def save_stl(verts, faces, filename):    
    data = mesh.Mesh(np.zeros(faces.shape[0], dtype=mesh.Mesh.dtype))
    for i, f in enumerate(faces):
        for j in range(3):
            data.vectors[i][j] = verts[f[j],:]
    data.save(filename)
    return data

save_stl(verts, faces, 'torch_tensor.stl')
```

The same code with `mcubes` works well. By the way I am opening stl files with blender.

**Comment by munrocket** [comment]
So it's a bug! Shape with 501x500x500 not supported.

**Comment by k-dominik** [comment]
Hi @munrocket,

so do you get an errror, or do you get an unexpected result from marching cubes? (or both?) is there a traceback in the case of an error?

**Comment by munrocket** [comment]
Yes! Library support only symmetrical shape. Check my data and script https://github.com/ilastik/marching_cubes/issues/25#issuecomment-663154076

**Comment by k-dominik** [comment]
could you please paste the error message. I do not get an error when doing marching cubes with your provided array.

**Comment by munrocket** [comment]
Now you can try to fix one line
```python
# quick fix for one line
verts, normals, faces = march(mask.narrow(0, 0, 500).numpy(), 4)
```
It fits to me for now, because I have all shapes with 501x500x500
As I understand this bug exists in your todo list. Thanks.

**Comment by DerThorsten** [comment]
I do not get an error message whatsoever with the provided volume:
```
import numpy as np
import torch

from stl import mesh
from marching_cubes import march

# import mcubes

mask = torch.load("/home/derthorsten/Downloads/torch_tensor.pt")

from pyqtgraph.opengl import GLViewWidget, MeshData
from pyqtgraph.opengl.items.GLMeshItem import GLMeshItem

from PyQt5.QtGui import QApplication


# extract the mesh where the values are larger than or equal to 1
# everything else is ignored
vertices, normals, faces = march(mask, 0)  # zero smoothing rounds
# smooth_vertices, smooth_normals, faces = march(volume, 4)  # 4 smoothing rounds


app = QApplication([])
view = GLViewWidget()

mesh = MeshData(
    vertices / 100, faces
)  # scale down - because camera is at a fixed position
# or mesh = MeshData(smooth_vertices / 100, faces)
mesh._vertexNormals = normals
# or mesh._vertexNormals = smooth_normals

item = GLMeshItem(meshdata=mesh, color=[1, 0, 0, 1], shader="normalColor")

view.addItem(item)
view.show()
app.exec_()
```

**Comment by DerThorsten** [comment]
But I can confirm that PyMcubes spits out  slightly different looking object (as in your sceenshot)

**Comment by DerThorsten** [comment]
It looks like the ikastikmarching cubes does smth wrong:
to debug I took a `GLVolumeItem` widget which can directly render a 3d pixel volume without marching cubes.
The `GLVoumenItem` result looks like the PyMCube results, while the ilastik marching cubes result looks different:
```python
import numpy as np
import torch

from stl import mesh
from marching_cubes import march

import mcubes

mask = torch.load("/home/derthorsten/Downloads/torch_tensor.pt")
mask = mask.data.numpy()

from pyqtgraph.opengl import GLViewWidget, MeshData, GLVolumeItem
from pyqtgraph.opengl.items.GLMeshItem import GLMeshItem

from PyQt5.QtGui import QApplication


vertices_ilastik, normals_ilastik, faces_ilastik = march(
    mask, 0
)  # zero smoothing rounds

vertices, faces = mcubes.marching_cubes(mask, 0)

# smooth_vertices, smooth_normals, faces = march(volume, 4)  # 4 smoothing rounds


app = QApplication([])
view = GLViewWidget()

# the mesh data
mesh_ilastik = MeshData(vertices_ilastik, faces_ilastik)
mesh = MeshData(vertices, faces)


item = GLMeshItem(meshdata=mesh, color=[1, 0, 0, 1], shader="normalColor")
item_ilastik = GLMeshItem(
    meshdata=mesh_ilastik, color=[0, 1, 0, 1], shader="normalColor"
)


# print(mask.min(), mask.max())

# as  volumetric
mask_4d = mask[..., None]
rgb_mask = np.repeat(mask_4d, repeats=4, axis=3)
rgb_mask = rgb_mask.astype("uint8") * 255
rgb_mask = np.array(rgb_mask)
volumetric = GLVolumeItem(rgb_mask)  # , smooth=False)

view.addItem(item)
view.addItem(item_ilastik)
view.addItem(volumetric)
view.setCameraPosition(distance=2000)
view.show()

app.exec_()
```

**Comment by DerThorsten** [comment]
When using a 500x500x500 shape (and making the mask F-Order for ilastik marching cubes) one gets the same results:
``` python
import numpy as np
import torch

from stl import mesh
from marching_cubes import march

import mcubes

mask = torch.load("/home/derthorsten/Downloads/torch_tensor.pt")
mask = mask.data.numpy()[0:500, 0:500, 0:500].copy()

from pyqtgraph.opengl import GLViewWidget, MeshData, GLVolumeItem
from pyqtgraph.opengl.items.GLMeshItem import GLMeshItem

from PyQt5.QtGui import QApplication


vertices_ilastik, normals_ilastik, faces_ilastik = march(
    np.require(mask, requirements=["F"]), 0
)  # zero smoothing rounds

vertices, faces = mcubes.marching_cubes(mask, 0)

# smooth_vertices, smooth_normals, faces = march(volume, 4)  # 4 smoothing rounds


app = QApplication([])
view = GLViewWidget()

# the mesh data
mesh_ilastik = MeshData(vertices_ilastik, faces_ilastik)
mesh = MeshData(vertices, faces)


item = GLMeshItem(meshdata=mesh, color=[1, 0, 0, 1], shader="normalColor")
item_ilastik = GLMeshItem(
    meshdata=mesh_ilastik, color=[0, 1, 0, 1], shader="normalColor"
)


# print(mask.min(), mask.max())

# as  volumetric
mask_4d = mask[..., None]
rgb_mask = np.repeat(mask_4d, repeats=4, axis=3)
rgb_mask = rgb_mask.astype("uint8") * 255
rgb_mask = np.array(rgb_mask)
volumetric = GLVolumeItem(rgb_mask)  # , smooth=False)

view.addItem(item)
view.addItem(item_ilastik)
view.addItem(volumetric)
view.setCameraPosition(distance=2000)
view.show()

app.exec_()
```

**Comment by DerThorsten** [comment]
update:
I think it is enough to just convert the array to F-Order  to get the correct results.
the error lies somewhere here:
https://github.com/ilastik/marching_cubes/blob/bfbeab71e9e4ff1396ce1b49476132f69a868a38/src/python_wrapper.cpp#L21 (missing f_style flag)
in conjunction with the shapes here
https://github.com/ilastik/marching_cubes/blob/bfbeab71e9e4ff1396ce1b49476132f69a868a38/src/python_wrapper.cpp#L32
And the manual index calculation everywhere (like here https://github.com/ilastik/marching_cubes/blob/bfbeab71e9e4ff1396ce1b49476132f69a868a38/src/marching_cubes.cpp#L334)

Maybe smth like this is enough: https://github.com/ilastik/marching_cubes/pull/26 (blind fix, nontested)

**Comment by k-dominik** [comment]
closed via #26

uploaded v0.3 builds for "early access" to `ilastik-forge/label/ilastik14` (will move to main soon)

so ... `conda create -n marching --override-channels -c ilastik-forge/label/ilastik14 -c conda-forge marching_cubes ... `
